### PR TITLE
adding download options to circulars

### LINKED ISSUE #1109: Tar file for Circulars archive
Alike classic GCN circulars, tar file of all GCN circulars would be useful feature.
We can add more option for this download option, e.x. date range?

### PR DESCRIPTION
**Pull request by Courey**
Fixes #1109.

Screen Shot 2023-08-02 at 10 32 03 AM

image

Adds the ability to download a single circular in json or text format.
Adds the ability to download entire database in json or text format as a tarball.

### CODE REVIEW

**Review by lpsinger — CHANGES_REQUESTED** [review]
- [x] Don't modify the UX on the individual Circulars page. (Those are deliberately in-browser links, not download links.)
- [x] Determine if we can stream the tarball rather than returning it as one big blob.
- [x] Format should be specified using routing, not search params.

**Review by lpsinger — CHANGES_REQUESTED** [review]
Please rebase. (Sorry, just merged #840, the big one.)

File: app/routes/circulars.download.json.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import { makeTarFile } from './circulars/circulars.server'
+
+export async function loader({ request }: DataFunctionArgs) {
+  const tarFile = await makeTarFile('json')
+
+  return new Response(tarFile, {
+    headers: {
+      'Content-Type': 'application/tar',
+      'Cache-Control': 'public, max-age=86400',
```
Comment (lpsinger) [inline app/routes/circulars.download.json.ts:18]: Please reuse one of the constants in headers.server.ts.

File: app/routes/circulars.download.txt.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// /*!
+//  * Copyright © 2022 United States Government as represented by the Administrator
+//  * of the National Aeronautics and Space Administration. No copyright is claimed
+//  * in the United States under Title 17, U.S. Code. All Other Rights Reserved.
+//  *
+//  * SPDX-License-Identifier: NASA-1.3
+//  */
```
Comment (lpsinger) [inline app/routes/circulars.download.txt.ts:7]:
```suggestion
/*!
 * Copyright © 2022 United States Government as represented by the Administrator
 * of the National Aeronautics and Space Administration. No copyright is claimed
 * in the United States under Title 17, U.S. Code. All Other Rights Reserved.
 *
 * SPDX-License-Identifier: NASA-1.3
 */
```

File: app/routes/circulars/circulars.server.ts [hunk 3]
```diff
@@ -263,3 +270,66 @@ export async function circularRedirect(query: string) {
...
+    }
+
+    const { Items, LastEvaluatedKey } = await data.circulars.scan(params)
+
+    if (Items) {
+      yield Items as Circular[]
+    }
+
+    exclusiveStartKey = LastEvaluatedKey
+  } while (exclusiveStartKey)
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:291]: Use a paginator rather than manually feeding the `LastEvaluatedKey` back to the API. There are several examples of scan paginators in our code that you can refer to.

File: package.json [hunk 4]
```diff
@@ -82,6 +83,7 @@
     "@testing-library/react": "^14.0.0",
     "@trivago/prettier-plugin-sort-imports": "^4.2.0",
     "@types/color-convert": "^2.0.0",
+    "@types/file-saver": "^2.0.5",
```
Comment (lpsinger) [inline package.json:86]: Unused dependency?

```suggestion
```

File: app/routes/circulars._index.tsx [hunk 5]
```diff
@@ -304,6 +312,67 @@ export default function () {
...
+                GCN Circulars Database Download
+              </ModalHeading>
+              <div className="usa-prose">
+                <div id="modal-1-description">
+                  This is a download of the entire GCN Circulars Databse.
+                </div>
+                <div>It may take a moment.</div>
+                <div>
+                  Download will start once a file format is chosen below.
+                </div>
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:341]: These should be paragraphs, not divs.

File: app/routes/circulars._index.tsx [hunk 6]
```diff
@@ -304,6 +312,67 @@ export default function () {
               )}
             </div>
           </div>
+          <div>
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:315]: Please move all of this into a component in this file.

File: app/routes/circulars._index.tsx [hunk 7]
```diff
@@ -276,7 +284,7 @@ export default function () {
               <div>
                 <Select
                   id="limit"
-                  className="width-auto height-5 padding-y-0 margin-y-0"
+                  className="width-auto height-5 padding-y-0 margin-bottom-2"
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:287]: What does this style change do?

File: app/routes/circulars._index.tsx [hunk 8]
```diff
@@ -304,6 +312,67 @@ export default function () {
...
             </div>
           </div>
+          <div>
+            <ModalToggleButton
+              modalRef={modalRef}
+              opener
+              className="text-middle"
+            >
+              <Icon.FileDownload className="bottom-aligned margin-right-05" />
+              Download
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:322]: This button needs a more descriptive label.

File: app/routes/circulars._index.tsx [hunk 9]
```diff
@@ -304,6 +312,67 @@ export default function () {
...
+                      download={'gcn-circulars.tar.gz'}
+                    >
+                      JSON
+                    </a>
+                  </ModalToggleButton>
+                  <ModalToggleButton
+                    modalRef={modalRef}
+                    closer
+                    unstyled
+                    className="padding-105 text-center"
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:368]: Why does this button require custom padding?

File: app/routes/circulars/circulars.server.ts [hunk 10]
```diff
@@ -263,3 +270,66 @@ export async function circularRedirect(query: string) {
...
+
+  return
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:293]: Is the return statement necessary?

File: app/routes/circulars/download.tsx [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  if (!fileType) return null
+  const tarFile = await makeTarFile(fileType)
+
+  return new Response(tarFile, {
+    headers: {
+      'Content-Type': 'application/tar',
+      'Cache-Control': 'public, max-age=86400',
+    },
+  })
+}
```
Comment (lpsinger) [inline app/routes/circulars/download.tsx:25]: This file appears to be unused.

File: app/routes/circulars._index.tsx [hunk 12]
```diff
@@ -304,6 +312,67 @@ export default function () {
...
+                      href={`/circulars/download/json`}
+                      download={'gcn-circulars.tar.gz'}
+                    >
+                      JSON
+                    </a>
+                  </ModalToggleButton>
+                  <ModalToggleButton
+                    modalRef={modalRef}
+                    closer
+                    unstyled
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:367]:
```suggestion
                    outline
```

File: app/routes/circulars._index.tsx [hunk 13]
```diff
@@ -304,6 +312,67 @@ export default function () {
...
+                    className="padding-105 text-center"
+                  >
+                    Go back
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:370]:
```suggestion
                    Cancel
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
- [x] Make sure to omit the `sub` field from the JSON serialized circulars. That information is sensitive. See https://github.com/nasa-gcn/gcn.nasa.gov/blob/386cf81a1e834cefd3be2d4b725f02a1a1482f3b/app/routes/circulars.%24circularId%5B.%5Djson.ts#L19
- [x] Suggest renaming the routes to /circulars/archive.json.tar and /circulars/archive.txt.tar so that they have a sensible default filenames when downloaded, even when not following the link from the modal.

File: app/routes/circulars._index.tsx (hunk 7 above)
Comment (Courey) [reply app/routes/circulars._index.tsx:287]: it didn't have any margin around the bottom so the modal button looked a bit crowded. It doesn't have to be there. I removed it. let me know if you feel like it's too crowded and I'll put it back in

**Review by lpsinger — CHANGES_REQUESTED** [review]
Please address the remaining unresolved issues:
- https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#pullrequestreview-1589537272
- https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1301704530
- https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1301708210
- https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1303264848
- https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1301718104

File: app/routes/circulars._index.tsx (hunk 7 above)
Comment (Courey) [reply app/routes/circulars._index.tsx]: Screen Shot 2023-08-23 at 10 33 02 AM
with padding:
Screen Shot 2023-08-23 at 10 33 34 AM

**Comment by Courey** [comment]
> Please address the remaining unresolved issues:
> 
>     * [adding download options to circulars #1182 (review)](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#pullrequestreview-1589537272)
> 
>     * [adding download options to circulars #1182 (comment)](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1301704530)
> 
>     * [adding download options to circulars #1182 (comment)](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1301708210)
> 
>     * [adding download options to circulars #1182 (comment)](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1303264848)
> 
>     * [adding download options to circulars #1182 (comment)](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1182#discussion_r1301718104)

I was still pushing changes. I'll hit the ready for re-review button when I'm ready for a re-review.

File: app/routes/circulars._index.tsx (hunk 7 above)
Comment (lpsinger) [reply app/routes/circulars._index.tsx]: Oh, yeah. You're right. With the adjusted padding, does it also look alright on mobile?

Could we add an "Advanced" heading down here?

Comment (Courey) [reply app/routes/circulars._index.tsx]: Screen Shot 2023-08-23 at 1 44 45 PM
Screen Shot 2023-08-23 at 1 48 26 PM

Comment (Courey) [reply app/routes/circulars._index.tsx]: and a third option:
Screen Shot 2023-08-23 at 1 50 04 PM
Screen Shot 2023-08-23 at 1 50 14 PM

Comment (lpsinger) [reply app/routes/circulars._index.tsx]: I like the version with a little bit more space. But can you add an "Advanced" heading and show what that looks like?

File: app/routes/circulars._index.tsx [hunk 14]
```diff
@@ -174,9 +180,78 @@ function Pagination({
...
+      <ModalToggleButton
+        modalRef={modalRef}
+        opener
+        outline
+        className="text-middle"
+      >
+        <Icon.FileDownload className="bottom-aligned margin-right-05" />
+        Download Archive
+      </ModalToggleButton>
+      <div className="margin-right-2"></div>
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:197]: Remove empty div?

```suggestion
```

File: app/routes/circulars._index.tsx [hunk 15]
```diff
@@ -174,9 +180,78 @@ function Pagination({
...
+        </ModalFooter>
+      </Modal>
+    </>
+  )
+}
+
 export default function () {
   const newItem = useActionData<typeof action>()
   const { items, page, totalPages, totalItems } = useLoaderData<typeof loader>()
+  const submit = useSubmit()
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:254]: Undo unnecessary moving around of the definitions of `submit` and `clean`.

File: app/routes/circulars._index.tsx [hunk 16]
```diff
@@ -174,9 +180,78 @@ function Pagination({
...
+          <p>It may take a moment.</p>
+          <p>Download will start once a file format is chosen below.</p>
+        </div>
+        <ModalFooter>
+          <ButtonGroup>
+            <ModalToggleButton modalRef={modalRef} closer>
+              <a
+                className="text-no-underline text-white"
+                href={`/circulars/archive/txt/tar`}
+                download={'gcn-circulars.tar.gz'}
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:221]:
```suggestion
                href={`/circulars/archive.txt.tar`}
```

File: app/routes/circulars._index.tsx [hunk 17]
```diff
@@ -174,9 +180,78 @@ function Pagination({
...
+              >
+                Raw text
+              </a>
+            </ModalToggleButton>
+
+            <ModalToggleButton modalRef={modalRef} closer>
+              <a
+                className="text-no-underline text-white"
+                href={`/circulars/archive/json/tar`}
+                download={'gcn-circulars.tar.gz'}
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:231]:
```suggestion
                href={`/circulars/archive.json.tar`}
```

File: app/routes/circulars/circulars.server.ts [hunk 18]
```diff
@@ -263,3 +270,57 @@ export async function circularRedirect(query: string) {
...
+  })
+
+  pack.on('end', () => {
+    tarStream.push(null)
+  })
+  for await (const circularArray of getAllRecords()) {
+    for (const circular of circularArray) {
+      if (fileType === 'txt') {
+        const txt_entry = pack.entry(
+          { name: `gcn-circulars/${circular.circularId}.txt` },
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:308]:
```suggestion
          { name: `archive/${circular.circularId}.txt` },
```

File: app/routes/circulars/circulars.server.ts [hunk 19]
```diff
@@ -263,3 +270,57 @@ export async function circularRedirect(query: string) {
...
+  const TableName = db.name('circulars')
+  const pages = paginateScan({ client }, { TableName })
+
+  for await (const page of pages) {
+    const items: Circular[] = page.Items as Circular[]
+    yield items
+  }
+}
+
+export async function makeTarFile(fileType: string): Promise<ReadableStream> {
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:286]:
```suggestion
export async function makeTarFile(fileType: 'json' | 'txt'): Promise<ReadableStream> {
```

File: app/routes/circulars._index.tsx [hunk 20]
```diff
@@ -174,9 +180,78 @@ function Pagination({
...
+        id="example-modal-1"
+        aria-labelledby="modal-1-heading"
+        aria-describedby="modal-1-description"
+      >
+        <ModalHeading id="modal-1-heading">
+          GCN Circulars Database Download
+        </ModalHeading>
+        <div className="usa-prose">
+          <p id="modal-1-description">
+            This is a download of the entire GCN Circulars Databse.
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:210]:
```suggestion
            This is a download of the entire GCN Circulars database.
```

File: app/routes/circulars._index.tsx [hunk 21]
```diff
@@ -174,9 +180,78 @@ function Pagination({
...
+          </p>
+          <p>It may take a moment.</p>
+          <p>Download will start once a file format is chosen below.</p>
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:213]: Fix passive voice.

File: app/routes/circulars._index.tsx [hunk 22]
```diff
@@ -39,6 +44,7 @@ export async function loader({ request: { url } }: DataFunctionArgs) {
   if (query) {
     await circularRedirect(query)
   }
+
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:47]: Undo unnecessary whitespace change.

```suggestion
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/circulars._index.tsx [hunk 23]
```diff
@@ -174,6 +179,73 @@ function Pagination({
...
+        id="example-modal-1"
+        aria-labelledby="modal-1-heading"
+        aria-describedby="modal-1-description"
+      >
+        <ModalHeading id="modal-1-heading">
+          GCN Circulars Database Download
+        </ModalHeading>
+        <div className="usa-prose">
+          <p id="modal-1-description">
+            This is a download of the entire GCN Circulars databse.
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:208]:
```suggestion
            This is a download of the entire GCN Circulars database.
```

File: app/routes/circulars._index.tsx [hunk 24]
```diff
@@ -174,6 +179,73 @@ function Pagination({
...
+          <p>It may take a moment.</p>
+          <p>Select a file format to begin download.</p>
+        </div>
+        <ModalFooter>
+          <ButtonGroup>
+            <ModalToggleButton modalRef={modalRef} closer>
+              <a
+                className="text-no-underline text-white"
+                href={`/circulars/archive.txt.tar`}
+                download={'archive.tar.gz'}
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:219]:
```suggestion
```

File: app/routes/circulars._index.tsx [hunk 25]
```diff
@@ -174,6 +179,73 @@ function Pagination({
...
+              >
+                Raw text
+              </a>
+            </ModalToggleButton>
+
+            <ModalToggleButton modalRef={modalRef} closer>
+              <a
+                className="text-no-underline text-white"
+                href={`/circulars/archive.json.tar`}
+                download={'archive.tar.gz'}
```
Comment (lpsinger) [inline app/routes/circulars._index.tsx:229]:
```suggestion
```

File: app/routes/circulars/circulars.server.ts [hunk 26]
```diff
@@ -263,3 +270,59 @@ export async function circularRedirect(query: string) {
...
+  })
+
+  pack.on('end', () => {
+    tarStream.push(null)
+  })
+  for await (const circularArray of getAllRecords()) {
+    for (const circular of circularArray) {
+      if (fileType === 'txt') {
+        const txt_entry = pack.entry(
+          { name: `archive/${circular.circularId}.txt` },
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:310]: The directory should match the name of the archive.

```suggestion
          { name: `archive.txt/${circular.circularId}.txt` },
```

File: app/routes/circulars/circulars.server.ts [hunk 27]
```diff
@@ -263,3 +270,59 @@ export async function circularRedirect(query: string) {
...
+          formatCircular(circular)
+        )
+        txt_entry.end()
+      } else if (fileType === 'json') {
+        delete circular.sub
+        const json_entry = pack.entry(
+          { name: `archive/${circular.circularId}.json` },
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:317]: The directory should match the name of the archive.

```suggestion
          { name: `archive.json/${circular.circularId}.json` },
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/circulars._index.tsx (hunk 24 above)
Comment (Courey) [reply app/routes/circulars._index.tsx:219]: Screen Shot 2023-08-24 at 9 19 38 AM
This is what happens if you take the download portion out of the a tag.

Comment (lpsinger) [reply app/routes/circulars._index.tsx:219]: Ah! That's because you aren't setting the MIME type of the response in your two new loaders. Please set the MIME type and then try removing the `download` attribute.

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — APPROVED** [review]
